From d817bf859fe944f8828efaf0a41313850a6253ac Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Tue, 28 Jan 2020 15:17:27 -0800
Subject: [PATCH 04/13] x86: Append missing ASM_END to asm files

Append missing ASM_END to x86 asm files to generate .note.gnu.property
section.

	* mpn/x86/aors_n.asm: Append ASM_END.
	* mpn/x86/aorsmul_1.asm: Likewise.
	* mpn/x86/atom/sse2/aorsmul_1.asm: Likewise.
	* mpn/x86/atom/sse2/mul_basecase.asm: Likewise.
	* mpn/x86/atom/sse2/sqr_basecase.asm: Likewise.
	* mpn/x86/bdiv_dbm1c.asm: Likewise.
	* mpn/x86/copyd.asm: Likewise.
	* mpn/x86/copyi.asm: Likewise.
	* mpn/x86/divrem_1.asm: Likewise.
	* mpn/x86/divrem_2.asm: Likewise.
	* mpn/x86/k6/aors_n.asm: Likewise.
	* mpn/x86/k6/aorsmul_1.asm: Likewise.
	* mpn/x86/k6/divrem_1.asm: Likewise.
	* mpn/x86/k6/k62mmx/copyd.asm: Likewise.
	* mpn/x86/k6/k62mmx/lshift.asm: Likewise.
	* mpn/x86/k6/k62mmx/rshift.asm: Likewise.
	* mpn/x86/k6/mmx/com.asm: Likewise.
	* mpn/x86/k6/mmx/logops_n.asm: Likewise.
	* mpn/x86/k6/mmx/lshift.asm: Likewise.
	* mpn/x86/k6/mmx/popham.asm: Likewise.
	* mpn/x86/k6/mmx/rshift.asm: Likewise.
	* mpn/x86/k6/mod_34lsub1.asm: Likewise.
	* mpn/x86/k6/mul_1.asm: Likewise.
	* mpn/x86/k6/mul_basecase.asm: Likewise.
	* mpn/x86/k6/pre_mod_1.asm: Likewise.
	* mpn/x86/k6/sqr_basecase.asm: Likewise.
	* mpn/x86/k7/aors_n.asm: Likewise.
	* mpn/x86/k7/mmx/com.asm: Likewise.
	* mpn/x86/k7/mmx/copyd.asm: Likewise.
	* mpn/x86/k7/mmx/copyi.asm: Likewise.
	* mpn/x86/k7/mmx/divrem_1.asm: Likewise.
	* mpn/x86/k7/mmx/lshift.asm: Likewise.
	* mpn/x86/k7/mmx/popham.asm: Likewise.
	* mpn/x86/k7/mmx/rshift.asm: Likewise.
	* mpn/x86/k7/mod_1_1.asm: Likewise.
	* mpn/x86/k7/mod_1_4.asm: Likewise.
	* mpn/x86/k7/mod_34lsub1.asm: Likewise.
	* mpn/x86/k7/mul_basecase.asm: Likewise.
	* mpn/x86/k7/sqr_basecase.asm: Likewise.
	* mpn/x86/lshift.asm: Likewise.
	* mpn/x86/mmx/sec_tabselect.asm: Likewise.
	* mpn/x86/mod_34lsub1.asm: Likewise.
	* mpn/x86/mul_1.asm: Likewise.
	* mpn/x86/mul_basecase.asm: Likewise.
	* mpn/x86/p6/aors_n.asm: Likewise.
	* mpn/x86/p6/aorsmul_1.asm: Likewise.
	* mpn/x86/p6/copyd.asm: Likewise.
	* mpn/x86/p6/gcd_11.asm: Likewise.
	* mpn/x86/p6/lshsub_n.asm: Likewise.
	* mpn/x86/p6/mmx/divrem_1.asm: Likewise.
	* mpn/x86/p6/mod_34lsub1.asm: Likewise.
	* mpn/x86/p6/mul_basecase.asm: Likewise.
	* mpn/x86/p6/sqr_basecase.asm: Likewise.
	* mpn/x86/pentium/aors_n.asm: Likewise.
	* mpn/x86/pentium/aorsmul_1.asm: Likewise.
	* mpn/x86/pentium/com.asm: Likewise.
	* mpn/x86/pentium/copyd.asm: Likewise.
	* mpn/x86/pentium/copyi.asm: Likewise.
	* mpn/x86/pentium/logops_n.asm: Likewise.
	* mpn/x86/pentium/lshift.asm: Likewise.
	* mpn/x86/pentium/mmx/lshift.asm: Likewise.
	* mpn/x86/pentium/mmx/mul_1.asm: Likewise.
	* mpn/x86/pentium/mmx/rshift.asm: Likewise.
	* mpn/x86/pentium/mod_34lsub1.asm: Likewise.
	* mpn/x86/pentium/mul_1.asm: Likewise.
	* mpn/x86/pentium/mul_2.asm: Likewise.
	* mpn/x86/pentium/mul_basecase.asm: Likewise.
	* mpn/x86/pentium/rshift.asm: Likewise.
	* mpn/x86/pentium/sqr_basecase.asm: Likewise.
	* mpn/x86/pentium4/copyd.asm: Likewise.
	* mpn/x86/pentium4/copyi.asm: Likewise.
	* mpn/x86/pentium4/mmx/popham.asm: Likewise.
	* mpn/x86/pentium4/sse2/add_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/addlsh1_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/addmul_1.asm: Likewise.
	* mpn/x86/pentium4/sse2/cnd_add_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/cnd_sub_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/divrem_1.asm: Likewise.
	* mpn/x86/pentium4/sse2/mod_1_1.asm: Likewise.
	* mpn/x86/pentium4/sse2/mod_1_4.asm: Likewise.
	* mpn/x86/pentium4/sse2/mod_34lsub1.asm: Likewise.
	* mpn/x86/pentium4/sse2/mul_1.asm: Likewise.
	* mpn/x86/pentium4/sse2/mul_basecase.asm: Likewise.
	* mpn/x86/pentium4/sse2/rsh1add_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/sqr_basecase.asm: Likewise.
	* mpn/x86/pentium4/sse2/sub_n.asm: Likewise.
	* mpn/x86/pentium4/sse2/submul_1.asm: Likewise.
	* mpn/x86/rshift.asm: Likewise.
	* mpn/x86/sec_tabselect.asm: Likewise.
	* mpn/x86/sqr_basecase.asm: Likewise.
	* mpn/x86/udiv.asm: Likewise.
	* mpn/x86/umul.asm: Likewise.
---
 mpn/x86/aors_n.asm                     | 1 +
 mpn/x86/aorsmul_1.asm                  | 1 +
 mpn/x86/atom/sse2/aorsmul_1.asm        | 1 +
 mpn/x86/atom/sse2/mul_basecase.asm     | 1 +
 mpn/x86/atom/sse2/sqr_basecase.asm     | 1 +
 mpn/x86/bdiv_dbm1c.asm                 | 1 +
 mpn/x86/copyd.asm                      | 1 +
 mpn/x86/copyi.asm                      | 1 +
 mpn/x86/divrem_1.asm                   | 1 +
 mpn/x86/divrem_2.asm                   | 1 +
 mpn/x86/k6/aors_n.asm                  | 1 +
 mpn/x86/k6/aorsmul_1.asm               | 1 +
 mpn/x86/k6/divrem_1.asm                | 1 +
 mpn/x86/k6/k62mmx/copyd.asm            | 1 +
 mpn/x86/k6/k62mmx/lshift.asm           | 1 +
 mpn/x86/k6/k62mmx/rshift.asm           | 1 +
 mpn/x86/k6/mmx/com.asm                 | 1 +
 mpn/x86/k6/mmx/logops_n.asm            | 1 +
 mpn/x86/k6/mmx/lshift.asm              | 1 +
 mpn/x86/k6/mmx/popham.asm              | 1 +
 mpn/x86/k6/mmx/rshift.asm              | 1 +
 mpn/x86/k6/mod_34lsub1.asm             | 1 +
 mpn/x86/k6/mul_1.asm                   | 1 +
 mpn/x86/k6/mul_basecase.asm            | 1 +
 mpn/x86/k6/pre_mod_1.asm               | 1 +
 mpn/x86/k6/sqr_basecase.asm            | 1 +
 mpn/x86/k7/aors_n.asm                  | 1 +
 mpn/x86/k7/mmx/com.asm                 | 1 +
 mpn/x86/k7/mmx/copyd.asm               | 1 +
 mpn/x86/k7/mmx/copyi.asm               | 1 +
 mpn/x86/k7/mmx/divrem_1.asm            | 1 +
 mpn/x86/k7/mmx/lshift.asm              | 1 +
 mpn/x86/k7/mmx/popham.asm              | 1 +
 mpn/x86/k7/mmx/rshift.asm              | 1 +
 mpn/x86/k7/mod_1_1.asm                 | 1 +
 mpn/x86/k7/mod_1_4.asm                 | 1 +
 mpn/x86/k7/mod_34lsub1.asm             | 1 +
 mpn/x86/k7/mul_basecase.asm            | 1 +
 mpn/x86/k7/sqr_basecase.asm            | 1 +
 mpn/x86/lshift.asm                     | 1 +
 mpn/x86/mmx/sec_tabselect.asm          | 1 +
 mpn/x86/mod_34lsub1.asm                | 1 +
 mpn/x86/mul_1.asm                      | 1 +
 mpn/x86/mul_basecase.asm               | 1 +
 mpn/x86/p6/aors_n.asm                  | 1 +
 mpn/x86/p6/aorsmul_1.asm               | 1 +
 mpn/x86/p6/copyd.asm                   | 1 +
 mpn/x86/p6/gcd_11.asm                  | 1 +
 mpn/x86/p6/lshsub_n.asm                | 1 +
 mpn/x86/p6/mmx/divrem_1.asm            | 1 +
 mpn/x86/p6/mod_34lsub1.asm             | 1 +
 mpn/x86/p6/mul_basecase.asm            | 1 +
 mpn/x86/p6/sqr_basecase.asm            | 1 +
 mpn/x86/pentium/aors_n.asm             | 1 +
 mpn/x86/pentium/aorsmul_1.asm          | 1 +
 mpn/x86/pentium/com.asm                | 1 +
 mpn/x86/pentium/copyd.asm              | 1 +
 mpn/x86/pentium/copyi.asm              | 1 +
 mpn/x86/pentium/logops_n.asm           | 1 +
 mpn/x86/pentium/lshift.asm             | 1 +
 mpn/x86/pentium/mmx/lshift.asm         | 1 +
 mpn/x86/pentium/mmx/mul_1.asm          | 1 +
 mpn/x86/pentium/mmx/rshift.asm         | 1 +
 mpn/x86/pentium/mod_34lsub1.asm        | 1 +
 mpn/x86/pentium/mul_1.asm              | 1 +
 mpn/x86/pentium/mul_2.asm              | 1 +
 mpn/x86/pentium/mul_basecase.asm       | 1 +
 mpn/x86/pentium/rshift.asm             | 1 +
 mpn/x86/pentium/sqr_basecase.asm       | 1 +
 mpn/x86/pentium4/copyd.asm             | 1 +
 mpn/x86/pentium4/copyi.asm             | 1 +
 mpn/x86/pentium4/mmx/popham.asm        | 1 +
 mpn/x86/pentium4/sse2/add_n.asm        | 1 +
 mpn/x86/pentium4/sse2/addlsh1_n.asm    | 1 +
 mpn/x86/pentium4/sse2/addmul_1.asm     | 1 +
 mpn/x86/pentium4/sse2/cnd_add_n.asm    | 1 +
 mpn/x86/pentium4/sse2/cnd_sub_n.asm    | 1 +
 mpn/x86/pentium4/sse2/divrem_1.asm     | 1 +
 mpn/x86/pentium4/sse2/mod_1_1.asm      | 1 +
 mpn/x86/pentium4/sse2/mod_1_4.asm      | 1 +
 mpn/x86/pentium4/sse2/mod_34lsub1.asm  | 1 +
 mpn/x86/pentium4/sse2/mul_1.asm        | 1 +
 mpn/x86/pentium4/sse2/mul_basecase.asm | 1 +
 mpn/x86/pentium4/sse2/rsh1add_n.asm    | 1 +
 mpn/x86/pentium4/sse2/sqr_basecase.asm | 1 +
 mpn/x86/pentium4/sse2/sub_n.asm        | 1 +
 mpn/x86/pentium4/sse2/submul_1.asm     | 1 +
 mpn/x86/rshift.asm                     | 1 +
 mpn/x86/sec_tabselect.asm              | 1 +
 mpn/x86/sqr_basecase.asm               | 1 +
 mpn/x86/udiv.asm                       | 1 +
 mpn/x86/umul.asm                       | 1 +
 92 files changed, 92 insertions(+)

diff --git a/mpn/x86/aors_n.asm b/mpn/x86/aors_n.asm
index 5d359f59b..95d006a5a 100644
--- a/mpn/x86/aors_n.asm
+++ b/mpn/x86/aors_n.asm
@@ -200,3 +200,4 @@ L(oop):	movl	(%esi),%eax
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/aorsmul_1.asm b/mpn/x86/aorsmul_1.asm
index 54a890544..0ab1e01f5 100644
--- a/mpn/x86/aorsmul_1.asm
+++ b/mpn/x86/aorsmul_1.asm
@@ -154,3 +154,4 @@ L(end):	movl	%ebx,%eax
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/atom/sse2/aorsmul_1.asm b/mpn/x86/atom/sse2/aorsmul_1.asm
index 969a14a91..20658e170 100644
--- a/mpn/x86/atom/sse2/aorsmul_1.asm
+++ b/mpn/x86/atom/sse2/aorsmul_1.asm
@@ -172,3 +172,4 @@ PROLOGUE(func_1c)
 	mov	20(%esp), %edx		C carry
 	jmp	L(ent)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/atom/sse2/mul_basecase.asm b/mpn/x86/atom/sse2/mul_basecase.asm
index 97d3aeb5a..74171aa5a 100644
--- a/mpn/x86/atom/sse2/mul_basecase.asm
+++ b/mpn/x86/atom/sse2/mul_basecase.asm
@@ -499,3 +499,4 @@ L(done):
 	pop	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/atom/sse2/sqr_basecase.asm b/mpn/x86/atom/sse2/sqr_basecase.asm
index af19ed854..0031812f6 100644
--- a/mpn/x86/atom/sse2/sqr_basecase.asm
+++ b/mpn/x86/atom/sse2/sqr_basecase.asm
@@ -632,3 +632,4 @@ L(one):	pmuludq	%mm7, %mm7
 	pop	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/bdiv_dbm1c.asm b/mpn/x86/bdiv_dbm1c.asm
index 0288c475c..7a3b1a62c 100644
--- a/mpn/x86/bdiv_dbm1c.asm
+++ b/mpn/x86/bdiv_dbm1c.asm
@@ -127,3 +127,4 @@ L(b1):	add	$-4, %ebp
 	pop	%esi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/copyd.asm b/mpn/x86/copyd.asm
index 51fa19568..0e588d997 100644
--- a/mpn/x86/copyd.asm
+++ b/mpn/x86/copyd.asm
@@ -89,3 +89,4 @@ PROLOGUE(mpn_copyd)
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/copyi.asm b/mpn/x86/copyi.asm
index f6b0354b4..6efbb903e 100644
--- a/mpn/x86/copyi.asm
+++ b/mpn/x86/copyi.asm
@@ -97,3 +97,4 @@ PROLOGUE(mpn_copyi)
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/divrem_1.asm b/mpn/x86/divrem_1.asm
index 255d4935c..b1af92042 100644
--- a/mpn/x86/divrem_1.asm
+++ b/mpn/x86/divrem_1.asm
@@ -231,3 +231,4 @@ deflit(`FRAME',8)
 	popl	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/divrem_2.asm b/mpn/x86/divrem_2.asm
index 4c38ad0ac..c2920c2d7 100644
--- a/mpn/x86/divrem_2.asm
+++ b/mpn/x86/divrem_2.asm
@@ -197,3 +197,4 @@ L(35):	sub	20(%esp), %ebp
 	movl	$1, 32(%esp)
 	jmp	L(8)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/aors_n.asm b/mpn/x86/k6/aors_n.asm
index 168f9b4ae..257ba5987 100644
--- a/mpn/x86/k6/aors_n.asm
+++ b/mpn/x86/k6/aors_n.asm
@@ -335,3 +335,4 @@ L(inplace_done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/aorsmul_1.asm b/mpn/x86/k6/aorsmul_1.asm
index eaa92ebb2..78be9d250 100644
--- a/mpn/x86/k6/aorsmul_1.asm
+++ b/mpn/x86/k6/aorsmul_1.asm
@@ -389,3 +389,4 @@ Zdisp(	M4_inst,%ecx, disp0,(%edi))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/divrem_1.asm b/mpn/x86/k6/divrem_1.asm
index b4cea4fa2..ca41a3f4d 100644
--- a/mpn/x86/k6/divrem_1.asm
+++ b/mpn/x86/k6/divrem_1.asm
@@ -201,3 +201,4 @@ deflit(`FRAME',8)
 	popl	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/k62mmx/copyd.asm b/mpn/x86/k6/k62mmx/copyd.asm
index f80a5a1cd..fc329f5b9 100644
--- a/mpn/x86/k6/k62mmx/copyd.asm
+++ b/mpn/x86/k6/k62mmx/copyd.asm
@@ -116,3 +116,4 @@ L(zero):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/k62mmx/lshift.asm b/mpn/x86/k6/k62mmx/lshift.asm
index c86575fee..728fb5b15 100644
--- a/mpn/x86/k6/k62mmx/lshift.asm
+++ b/mpn/x86/k6/k62mmx/lshift.asm
@@ -292,3 +292,4 @@ deflit(`FRAME',4)
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/k62mmx/rshift.asm b/mpn/x86/k6/k62mmx/rshift.asm
index f604a7bd5..bd673f303 100644
--- a/mpn/x86/k6/k62mmx/rshift.asm
+++ b/mpn/x86/k6/k62mmx/rshift.asm
@@ -291,3 +291,4 @@ L(finish_even):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mmx/com.asm b/mpn/x86/k6/mmx/com.asm
index b74745462..646d16b02 100644
--- a/mpn/x86/k6/mmx/com.asm
+++ b/mpn/x86/k6/mmx/com.asm
@@ -101,3 +101,4 @@ L(no_extra):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mmx/logops_n.asm b/mpn/x86/k6/mmx/logops_n.asm
index e17930bb2..acfd7df08 100644
--- a/mpn/x86/k6/mmx/logops_n.asm
+++ b/mpn/x86/k6/mmx/logops_n.asm
@@ -224,3 +224,4 @@ L(no_extra):
 			ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mmx/lshift.asm b/mpn/x86/k6/mmx/lshift.asm
index 45be58263..eee1eb869 100644
--- a/mpn/x86/k6/mmx/lshift.asm
+++ b/mpn/x86/k6/mmx/lshift.asm
@@ -128,3 +128,4 @@ L(top):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mmx/popham.asm b/mpn/x86/k6/mmx/popham.asm
index 2b19d0b5e..efeb1b4cb 100644
--- a/mpn/x86/k6/mmx/popham.asm
+++ b/mpn/x86/k6/mmx/popham.asm
@@ -234,3 +234,4 @@ HAM(`	nop			C code alignment')
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mmx/rshift.asm b/mpn/x86/k6/mmx/rshift.asm
index cd0382f32..ae53711cb 100644
--- a/mpn/x86/k6/mmx/rshift.asm
+++ b/mpn/x86/k6/mmx/rshift.asm
@@ -128,3 +128,4 @@ Zdisp(	movd,	%mm0, 0,(%ecx,%eax,4))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mod_34lsub1.asm b/mpn/x86/k6/mod_34lsub1.asm
index 7e30503e5..05f89795c 100644
--- a/mpn/x86/k6/mod_34lsub1.asm
+++ b/mpn/x86/k6/mod_34lsub1.asm
@@ -188,3 +188,4 @@ L(combine):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mul_1.asm b/mpn/x86/k6/mul_1.asm
index 3ef7ec24f..2139f3673 100644
--- a/mpn/x86/k6/mul_1.asm
+++ b/mpn/x86/k6/mul_1.asm
@@ -290,3 +290,4 @@ L(finish_not_one):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/mul_basecase.asm b/mpn/x86/k6/mul_basecase.asm
index 7030001c3..ab202a2ce 100644
--- a/mpn/x86/k6/mul_basecase.asm
+++ b/mpn/x86/k6/mul_basecase.asm
@@ -610,3 +610,4 @@ Zdisp(	addl,	%ecx, disp0,(%edi))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/pre_mod_1.asm b/mpn/x86/k6/pre_mod_1.asm
index 34db20d38..1e4cb17e9 100644
--- a/mpn/x86/k6/pre_mod_1.asm
+++ b/mpn/x86/k6/pre_mod_1.asm
@@ -144,3 +144,4 @@ L(q1_ff):
 
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k6/sqr_basecase.asm b/mpn/x86/k6/sqr_basecase.asm
index b7ecb5cc8..f3a101a0f 100644
--- a/mpn/x86/k6/sqr_basecase.asm
+++ b/mpn/x86/k6/sqr_basecase.asm
@@ -678,3 +678,4 @@ L(pic_calc):
 
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/aors_n.asm b/mpn/x86/k7/aors_n.asm
index 1a0807202..bfdf3d40a 100644
--- a/mpn/x86/k7/aors_n.asm
+++ b/mpn/x86/k7/aors_n.asm
@@ -256,3 +256,4 @@ L(even):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/com.asm b/mpn/x86/k7/mmx/com.asm
index a258c224f..cf48facfd 100644
--- a/mpn/x86/k7/mmx/com.asm
+++ b/mpn/x86/k7/mmx/com.asm
@@ -123,3 +123,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/copyd.asm b/mpn/x86/k7/mmx/copyd.asm
index 59ece4092..3bc9ff8b8 100644
--- a/mpn/x86/k7/mmx/copyd.asm
+++ b/mpn/x86/k7/mmx/copyd.asm
@@ -142,3 +142,4 @@ L(done):
 
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/copyi.asm b/mpn/x86/k7/mmx/copyi.asm
index 9a28f927e..f0648fa09 100644
--- a/mpn/x86/k7/mmx/copyi.asm
+++ b/mpn/x86/k7/mmx/copyi.asm
@@ -155,3 +155,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/divrem_1.asm b/mpn/x86/k7/mmx/divrem_1.asm
index cf343280b..370bfbb88 100644
--- a/mpn/x86/k7/mmx/divrem_1.asm
+++ b/mpn/x86/k7/mmx/divrem_1.asm
@@ -830,3 +830,4 @@ L(fraction_entry):
 	jmp	L(fraction_done)
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/lshift.asm b/mpn/x86/k7/mmx/lshift.asm
index b3383cf2c..4140e82bb 100644
--- a/mpn/x86/k7/mmx/lshift.asm
+++ b/mpn/x86/k7/mmx/lshift.asm
@@ -479,3 +479,4 @@ L(end_even_unaligned):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/popham.asm b/mpn/x86/k7/mmx/popham.asm
index 95965b74d..f29540a0a 100644
--- a/mpn/x86/k7/mmx/popham.asm
+++ b/mpn/x86/k7/mmx/popham.asm
@@ -211,3 +211,4 @@ L(loaded):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mmx/rshift.asm b/mpn/x86/k7/mmx/rshift.asm
index 345d23a25..0da1f9341 100644
--- a/mpn/x86/k7/mmx/rshift.asm
+++ b/mpn/x86/k7/mmx/rshift.asm
@@ -478,3 +478,4 @@ L(end_even_unaligned):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mod_1_1.asm b/mpn/x86/k7/mod_1_1.asm
index 1bbe6f92d..8da9519b0 100644
--- a/mpn/x86/k7/mod_1_1.asm
+++ b/mpn/x86/k7/mod_1_1.asm
@@ -219,3 +219,4 @@ PROLOGUE(mpn_mod_1_1p_cps)
 	pop	%ebp
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mod_1_4.asm b/mpn/x86/k7/mod_1_4.asm
index bb7597edd..fe1da5bba 100644
--- a/mpn/x86/k7/mod_1_4.asm
+++ b/mpn/x86/k7/mod_1_4.asm
@@ -258,3 +258,4 @@ C CAUTION: This is the same code as in pentium4/sse2/mod_1_4.asm
 	pop	%ebp
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mod_34lsub1.asm b/mpn/x86/k7/mod_34lsub1.asm
index ee3ad0409..0c1b8c84e 100644
--- a/mpn/x86/k7/mod_34lsub1.asm
+++ b/mpn/x86/k7/mod_34lsub1.asm
@@ -186,3 +186,4 @@ L(combine):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/mul_basecase.asm b/mpn/x86/k7/mul_basecase.asm
index 4dfb50088..b96fda7e0 100644
--- a/mpn/x86/k7/mul_basecase.asm
+++ b/mpn/x86/k7/mul_basecase.asm
@@ -600,3 +600,4 @@ deflit(`disp1', eval(disp0-0 + 4))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/k7/sqr_basecase.asm b/mpn/x86/k7/sqr_basecase.asm
index 7b6a97e0d..df47ee421 100644
--- a/mpn/x86/k7/sqr_basecase.asm
+++ b/mpn/x86/k7/sqr_basecase.asm
@@ -633,3 +633,4 @@ L(diag):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/lshift.asm b/mpn/x86/lshift.asm
index 6ee6153cc..95f5321e0 100644
--- a/mpn/x86/lshift.asm
+++ b/mpn/x86/lshift.asm
@@ -104,3 +104,4 @@ L(end):	shll	%cl,%ebx		C compute least significant limb
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/mmx/sec_tabselect.asm b/mpn/x86/mmx/sec_tabselect.asm
index aae158abf..543dec164 100644
--- a/mpn/x86/mmx/sec_tabselect.asm
+++ b/mpn/x86/mmx/sec_tabselect.asm
@@ -161,3 +161,4 @@ L(b00):	pop	%ebp
 	emms
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/mod_34lsub1.asm b/mpn/x86/mod_34lsub1.asm
index e09e702c6..df52d371c 100644
--- a/mpn/x86/mod_34lsub1.asm
+++ b/mpn/x86/mod_34lsub1.asm
@@ -181,3 +181,4 @@ L(combine):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/mul_1.asm b/mpn/x86/mul_1.asm
index 421de6222..dbbc0e339 100644
--- a/mpn/x86/mul_1.asm
+++ b/mpn/x86/mul_1.asm
@@ -138,3 +138,4 @@ L(end):	movl	%ebx,%eax
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/mul_basecase.asm b/mpn/x86/mul_basecase.asm
index 8339732a8..c32fd7e0e 100644
--- a/mpn/x86/mul_basecase.asm
+++ b/mpn/x86/mul_basecase.asm
@@ -221,3 +221,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/aors_n.asm b/mpn/x86/p6/aors_n.asm
index df51c2e6f..b4a32150f 100644
--- a/mpn/x86/p6/aors_n.asm
+++ b/mpn/x86/p6/aors_n.asm
@@ -154,3 +154,4 @@ PROLOGUE(func_nc)
 	movl	20(%esp), %edx
 	jmp	L(start)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/aorsmul_1.asm b/mpn/x86/p6/aorsmul_1.asm
index bc8c49c62..d6bc549e3 100644
--- a/mpn/x86/p6/aorsmul_1.asm
+++ b/mpn/x86/p6/aorsmul_1.asm
@@ -318,3 +318,4 @@ deflit(`disp0',	eval(UNROLL_BYTES ifelse(UNROLL_BYTES,256,-128)))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/copyd.asm b/mpn/x86/p6/copyd.asm
index 1be763683..bd42da18b 100644
--- a/mpn/x86/p6/copyd.asm
+++ b/mpn/x86/p6/copyd.asm
@@ -176,3 +176,4 @@ L(zero):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/gcd_11.asm b/mpn/x86/p6/gcd_11.asm
index 80e055e2e..a7fc6a8d6 100644
--- a/mpn/x86/p6/gcd_11.asm
+++ b/mpn/x86/p6/gcd_11.asm
@@ -81,3 +81,4 @@ L(end):	mov	%edx, %eax
 	pop	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/lshsub_n.asm b/mpn/x86/p6/lshsub_n.asm
index 7ada21364..2ceb98bd8 100644
--- a/mpn/x86/p6/lshsub_n.asm
+++ b/mpn/x86/p6/lshsub_n.asm
@@ -167,3 +167,4 @@ L(ent):	mov	   0(up,n,4), %eax
 	jmp	   L(top)
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/mmx/divrem_1.asm b/mpn/x86/p6/mmx/divrem_1.asm
index 5300616c1..b6057ddcc 100644
--- a/mpn/x86/p6/mmx/divrem_1.asm
+++ b/mpn/x86/p6/mmx/divrem_1.asm
@@ -765,3 +765,4 @@ L(fraction_top):
 	jmp	L(fraction_done)
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/mod_34lsub1.asm b/mpn/x86/p6/mod_34lsub1.asm
index b88ab5d17..46b38060d 100644
--- a/mpn/x86/p6/mod_34lsub1.asm
+++ b/mpn/x86/p6/mod_34lsub1.asm
@@ -188,3 +188,4 @@ L(done_0):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/mul_basecase.asm b/mpn/x86/p6/mul_basecase.asm
index d87bc12b6..03ae0d6b2 100644
--- a/mpn/x86/p6/mul_basecase.asm
+++ b/mpn/x86/p6/mul_basecase.asm
@@ -605,3 +605,4 @@ deflit(`disp1', eval(disp0 + 4))
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/p6/sqr_basecase.asm b/mpn/x86/p6/sqr_basecase.asm
index 8fc7fdf37..ca29e8084 100644
--- a/mpn/x86/p6/sqr_basecase.asm
+++ b/mpn/x86/p6/sqr_basecase.asm
@@ -647,3 +647,4 @@ L(pic_calc):
 
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/aors_n.asm b/mpn/x86/pentium/aors_n.asm
index 01ebfb96a..ca124a54f 100644
--- a/mpn/x86/pentium/aors_n.asm
+++ b/mpn/x86/pentium/aors_n.asm
@@ -201,3 +201,4 @@ L(end2):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/aorsmul_1.asm b/mpn/x86/pentium/aorsmul_1.asm
index d83cc4513..5cec8b3ad 100644
--- a/mpn/x86/pentium/aorsmul_1.asm
+++ b/mpn/x86/pentium/aorsmul_1.asm
@@ -142,3 +142,4 @@ L(top):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/com.asm b/mpn/x86/pentium/com.asm
index b0805452a..00064fffe 100644
--- a/mpn/x86/pentium/com.asm
+++ b/mpn/x86/pentium/com.asm
@@ -179,3 +179,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/copyd.asm b/mpn/x86/pentium/copyd.asm
index 72a543b2a..c7f74b5fc 100644
--- a/mpn/x86/pentium/copyd.asm
+++ b/mpn/x86/pentium/copyd.asm
@@ -144,3 +144,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/copyi.asm b/mpn/x86/pentium/copyi.asm
index d983d6b46..bc7744eb8 100644
--- a/mpn/x86/pentium/copyi.asm
+++ b/mpn/x86/pentium/copyi.asm
@@ -162,3 +162,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/logops_n.asm b/mpn/x86/pentium/logops_n.asm
index 18773172e..41a9477ba 100644
--- a/mpn/x86/pentium/logops_n.asm
+++ b/mpn/x86/pentium/logops_n.asm
@@ -174,3 +174,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/lshift.asm b/mpn/x86/pentium/lshift.asm
index 2a31f36c6..68cba52e9 100644
--- a/mpn/x86/pentium/lshift.asm
+++ b/mpn/x86/pentium/lshift.asm
@@ -241,3 +241,4 @@ L(L1):	movl	%edx,(%edi)		C store last limb
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mmx/lshift.asm b/mpn/x86/pentium/mmx/lshift.asm
index 04b0ddcc8..9e18c861e 100644
--- a/mpn/x86/pentium/mmx/lshift.asm
+++ b/mpn/x86/pentium/mmx/lshift.asm
@@ -461,3 +461,4 @@ L(finish_zero_unaligned):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mmx/mul_1.asm b/mpn/x86/pentium/mmx/mul_1.asm
index 4ced577b1..b04a718f8 100644
--- a/mpn/x86/pentium/mmx/mul_1.asm
+++ b/mpn/x86/pentium/mmx/mul_1.asm
@@ -369,3 +369,4 @@ L(small_done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mmx/rshift.asm b/mpn/x86/pentium/mmx/rshift.asm
index e3b274bb6..5493d2054 100644
--- a/mpn/x86/pentium/mmx/rshift.asm
+++ b/mpn/x86/pentium/mmx/rshift.asm
@@ -466,3 +466,4 @@ L(finish_zero_unaligned):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mod_34lsub1.asm b/mpn/x86/pentium/mod_34lsub1.asm
index 2d88223b8..0945de825 100644
--- a/mpn/x86/pentium/mod_34lsub1.asm
+++ b/mpn/x86/pentium/mod_34lsub1.asm
@@ -190,3 +190,4 @@ L(combine):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mul_1.asm b/mpn/x86/pentium/mul_1.asm
index a0858af2b..2c4913071 100644
--- a/mpn/x86/pentium/mul_1.asm
+++ b/mpn/x86/pentium/mul_1.asm
@@ -175,3 +175,4 @@ L(top):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mul_2.asm b/mpn/x86/pentium/mul_2.asm
index 4c7beb5df..e94e07120 100644
--- a/mpn/x86/pentium/mul_2.asm
+++ b/mpn/x86/pentium/mul_2.asm
@@ -148,3 +148,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/mul_basecase.asm b/mpn/x86/pentium/mul_basecase.asm
index e1d0f0526..ff269bb9d 100644
--- a/mpn/x86/pentium/mul_basecase.asm
+++ b/mpn/x86/pentium/mul_basecase.asm
@@ -140,3 +140,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/rshift.asm b/mpn/x86/pentium/rshift.asm
index 2105c4c93..d98080de6 100644
--- a/mpn/x86/pentium/rshift.asm
+++ b/mpn/x86/pentium/rshift.asm
@@ -241,3 +241,4 @@ L(L1):	movl	%edx,(%edi)		C store last limb
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium/sqr_basecase.asm b/mpn/x86/pentium/sqr_basecase.asm
index b11d767da..ee64eb371 100644
--- a/mpn/x86/pentium/sqr_basecase.asm
+++ b/mpn/x86/pentium/sqr_basecase.asm
@@ -526,3 +526,4 @@ L(diag):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/copyd.asm b/mpn/x86/pentium4/copyd.asm
index 82af81c52..bf06a0534 100644
--- a/mpn/x86/pentium4/copyd.asm
+++ b/mpn/x86/pentium4/copyd.asm
@@ -69,3 +69,4 @@ L(end):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/copyi.asm b/mpn/x86/pentium4/copyi.asm
index b6148879f..acbb3f462 100644
--- a/mpn/x86/pentium4/copyi.asm
+++ b/mpn/x86/pentium4/copyi.asm
@@ -91,3 +91,4 @@ L(replmovs):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/mmx/popham.asm b/mpn/x86/pentium4/mmx/popham.asm
index 9563cb57e..f7a612487 100644
--- a/mpn/x86/pentium4/mmx/popham.asm
+++ b/mpn/x86/pentium4/mmx/popham.asm
@@ -201,3 +201,4 @@ L(loaded):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/add_n.asm b/mpn/x86/pentium4/sse2/add_n.asm
index 8e2380e49..e32963560 100644
--- a/mpn/x86/pentium4/sse2/add_n.asm
+++ b/mpn/x86/pentium4/sse2/add_n.asm
@@ -99,3 +99,4 @@ L(top):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/addlsh1_n.asm b/mpn/x86/pentium4/sse2/addlsh1_n.asm
index 93b63b201..e801f7b4f 100644
--- a/mpn/x86/pentium4/sse2/addlsh1_n.asm
+++ b/mpn/x86/pentium4/sse2/addlsh1_n.asm
@@ -106,3 +106,4 @@ L(top):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/addmul_1.asm b/mpn/x86/pentium4/sse2/addmul_1.asm
index 78102072b..62a76753a 100644
--- a/mpn/x86/pentium4/sse2/addmul_1.asm
+++ b/mpn/x86/pentium4/sse2/addmul_1.asm
@@ -187,3 +187,4 @@ PROLOGUE(mpn_addmul_1c)
 	movd	20(%esp), %mm6
 	jmp	L(ent)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/cnd_add_n.asm b/mpn/x86/pentium4/sse2/cnd_add_n.asm
index b3f3474e6..7183b9422 100644
--- a/mpn/x86/pentium4/sse2/cnd_add_n.asm
+++ b/mpn/x86/pentium4/sse2/cnd_add_n.asm
@@ -93,3 +93,4 @@ L(top):	movd	(%ebx,%ecx,4), %mm2
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/cnd_sub_n.asm b/mpn/x86/pentium4/sse2/cnd_sub_n.asm
index 339a23e0b..ba0fc47ad 100644
--- a/mpn/x86/pentium4/sse2/cnd_sub_n.asm
+++ b/mpn/x86/pentium4/sse2/cnd_sub_n.asm
@@ -112,3 +112,4 @@ L(done_mm1):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/divrem_1.asm b/mpn/x86/pentium4/sse2/divrem_1.asm
index 0146fab11..d8619e047 100644
--- a/mpn/x86/pentium4/sse2/divrem_1.asm
+++ b/mpn/x86/pentium4/sse2/divrem_1.asm
@@ -643,3 +643,4 @@ L(fraction_top):
 	jmp	L(fraction_done)
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/mod_1_1.asm b/mpn/x86/pentium4/sse2/mod_1_1.asm
index ee88babee..2e5a5146b 100644
--- a/mpn/x86/pentium4/sse2/mod_1_1.asm
+++ b/mpn/x86/pentium4/sse2/mod_1_1.asm
@@ -164,3 +164,4 @@ C CAUTION: This is the same code as in k7/mod_1_1.asm
 	pop	%ebp
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/mod_1_4.asm b/mpn/x86/pentium4/sse2/mod_1_4.asm
index eb2edb629..5ef3c4aef 100644
--- a/mpn/x86/pentium4/sse2/mod_1_4.asm
+++ b/mpn/x86/pentium4/sse2/mod_1_4.asm
@@ -267,3 +267,4 @@ C CAUTION: This is the same code as in k7/mod_1_4.asm
 	pop	%ebp
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/mod_34lsub1.asm b/mpn/x86/pentium4/sse2/mod_34lsub1.asm
index 31e25b79b..5b6b9a765 100644
--- a/mpn/x86/pentium4/sse2/mod_34lsub1.asm
+++ b/mpn/x86/pentium4/sse2/mod_34lsub1.asm
@@ -173,3 +173,4 @@ L(combine):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/mul_1.asm b/mpn/x86/pentium4/sse2/mul_1.asm
index 6347b8bf6..9e4f3fc88 100644
--- a/mpn/x86/pentium4/sse2/mul_1.asm
+++ b/mpn/x86/pentium4/sse2/mul_1.asm
@@ -162,3 +162,4 @@ PROLOGUE(mpn_mul_1c)
 	movd	20(%esp), %mm6
 	jmp	L(ent)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/mul_basecase.asm b/mpn/x86/pentium4/sse2/mul_basecase.asm
index 6e3775ae0..0bad75610 100644
--- a/mpn/x86/pentium4/sse2/mul_basecase.asm
+++ b/mpn/x86/pentium4/sse2/mul_basecase.asm
@@ -660,3 +660,4 @@ L(oel3):
 	pop	%esi			C				   3
 	ret				C				   3
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/rsh1add_n.asm b/mpn/x86/pentium4/sse2/rsh1add_n.asm
index f421d1323..543a637de 100644
--- a/mpn/x86/pentium4/sse2/rsh1add_n.asm
+++ b/mpn/x86/pentium4/sse2/rsh1add_n.asm
@@ -124,3 +124,4 @@ L(done):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/sqr_basecase.asm b/mpn/x86/pentium4/sse2/sqr_basecase.asm
index 2dd57d25d..9695d42c5 100644
--- a/mpn/x86/pentium4/sse2/sqr_basecase.asm
+++ b/mpn/x86/pentium4/sse2/sqr_basecase.asm
@@ -703,3 +703,4 @@ L(diag):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/sub_n.asm b/mpn/x86/pentium4/sse2/sub_n.asm
index 5ba1c018e..2cd5b2297 100644
--- a/mpn/x86/pentium4/sse2/sub_n.asm
+++ b/mpn/x86/pentium4/sse2/sub_n.asm
@@ -117,3 +117,4 @@ L(done_mm1):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/pentium4/sse2/submul_1.asm b/mpn/x86/pentium4/sse2/submul_1.asm
index 020675bd7..1172f0a5a 100644
--- a/mpn/x86/pentium4/sse2/submul_1.asm
+++ b/mpn/x86/pentium4/sse2/submul_1.asm
@@ -180,3 +180,4 @@ L(eod):	paddq	%mm6, %mm4		C add 0xFFFFFFFE00000001
 	movd	%mm0, 8(%edx)		C result
 	jmp	L(rt)
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/rshift.asm b/mpn/x86/rshift.asm
index a60dcaa4b..1cedc0ddc 100644
--- a/mpn/x86/rshift.asm
+++ b/mpn/x86/rshift.asm
@@ -106,3 +106,4 @@ L(end):	shrl	%cl,%ebx		C compute most significant limb
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/sec_tabselect.asm b/mpn/x86/sec_tabselect.asm
index c7c2e059f..3a8fa1729 100644
--- a/mpn/x86/sec_tabselect.asm
+++ b/mpn/x86/sec_tabselect.asm
@@ -113,3 +113,4 @@ L(outer_end):
 	pop	%edi
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/sqr_basecase.asm b/mpn/x86/sqr_basecase.asm
index 39f8a8980..3414b057c 100644
--- a/mpn/x86/sqr_basecase.asm
+++ b/mpn/x86/sqr_basecase.asm
@@ -357,3 +357,4 @@ L(diag):
 	ret
 
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/udiv.asm b/mpn/x86/udiv.asm
index a3ee08860..2531ef72c 100644
--- a/mpn/x86/udiv.asm
+++ b/mpn/x86/udiv.asm
@@ -50,3 +50,4 @@ deflit(`FRAME',0)
 	movl	%edx, (%ecx)
 	ret
 EPILOGUE()
+ASM_END()
diff --git a/mpn/x86/umul.asm b/mpn/x86/umul.asm
index 34fe43440..5c1da35c9 100644
--- a/mpn/x86/umul.asm
+++ b/mpn/x86/umul.asm
@@ -49,3 +49,4 @@ deflit(`FRAME',0)
 	movl	%edx, %eax
 	ret
 EPILOGUE()
+ASM_END()
-- 
2.27.0

